spec/foundation.md

```diff
@@ -0,0 +1,615 @@
+By binding such operations to the lifetime of client IDs, the shell can reset the terminal into a known state when a child process crashes, by simply ending the lifetime of the crashed process's client ID.
+
+There need not be a separate process for each client ID.
+A client can also choose new client IDs for itself, to describe durations inside the lifetime of its main client ID.
+
+## 3. Messages
+
+### 3.1. Syntax
+
+#### 3.1.1. Netstrings
+
+```abnf
+byte   = %x00-FF ; any single byte
+length = "0" / nzdigit *digit
+
+netstring = length ":" *byte ","
+```
+
+A **netstring** is a sequence of bytes such that:
+
+- it is accepted by `<netstring>`, and
+- the number of bytes between the first colon and the trailing comma is equal to the decimal value of the sequence of ASCII digits before the first colon.
+
+The sequence of bytes between the first colon and the trailing comma is called the **value** of the netstring.
+
+*Rationale:* The [netstring encoding](https://cr.yp.to/proto/netstrings.txt) was first described by Daniel J. Bernstein.
+
+#### 3.1.2. Messages
+
+```abnf
+message-without-client-id = "{" length "|" 1*netstring "}"
+message-with-client-id    = "{" netstring length "|" 1*netstring "}"
+```
+
+A **VT6 message** (or just **message**, if the term is not ambiguous) is a bytestring such that:
+
+- it is accepted by `<message-without-client-id>` or `<message-with-client-id>`,
+- the number of netstrings in the sequence after the pipe symbol is equal to the decimal value of the sequence of ASCII digits before the pipe symbol, and
+- the value of the first netstring after the pipe symbol is accepted by `<message-type>`, and
+- the entire message, including the surrounding braces, does not exceed a length of 1024 bytes.
+
+Furthermore, for messages accepted by `<message-with-client-id>`, the value of the netstring before the pipe symbol MUST be accepted by `<client-id>`.
+
+The first netstring after the pipe symbol is called the message's **type**.
+Any following netstrings are called the message's **arguments**.
+A message may have arbitrarily many arguments, including zero arguments.
+
+*Rationale:* We avoid compactly-coded escape sequences like those specified by [ECMA-48](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-48,%202nd%20Edition,%20August%201979.pdf) aka ANSI&nbsp;X3.64 aka ISO/IEC&nbsp;6429 because of the risk that escape sequences specified by different modules collide with each other.
+We choose an adaptation of the netstring format because it can be implemented very easily, both on the generating and on the parsing side.
+The curly brackets that enclose messages serve as a sequence point where parsing can be restarted after a parsing error (see sections 2.3 and 2.4).
+The maximum size limit of 1024 bytes reduces implementation complexity for simple client processes that cannot perform (or do not want to perform) dynamic memory allocation.
```

Comment:
May increase implementation complexity for the processes that generate those messages though, because they have to check on the length each time they generate a message.